### PR TITLE
Alert unit email and us if storage threshold reached

### PR DESCRIPTION
**Pull request by rv0lt**
## Read this before submitting the PR

1. Always create a Draft PR first
2. Go through sections 1-5 below, fill them in and check all the boxes
3. Make sure that the branch is updated; if there's an "Update branch" button at the bottom of the PR, rebase or update branch.
4. When all boxes are checked, information is filled in, and the branch is updated: mark as Ready For Review and tag reviewers (top right)
5. Once there is a submitted review, implement the suggestions (if reasonable, otherwise discuss) and request an new review.

If there is a field which you are unsure about, enter the edit mode of this description or go to the [PR template](../.github/pull_request_template.md); There are invisible comments providing descriptions which may be of help.

## 1. Description / Summary

Modify the monitor usage command to send an email to both the unit affected and our team in case the warning level is surpassed

## 2. Jira task / GitHub issue

[_Link to the github issue or add the Jira task ID here._](https://scilifelab.atlassian.net/jira/software/projects/DDS/boards/13?selectedIssue=DDS-1675)

## 3. Type of change

What _type of change(s)_ does the PR contain?

**Check the relevant boxes below. For an explanation of the different sections, enter edit mode of this PR description template.**

- [x] New feature
  - [ ] Breaking: _Why / How? Add info here._ 
  - [x] Non-breaking 
- [x] Database change: _Remember the to include a new migration version, **or** explain here why it's not needed._ 
- [ ] Bug fix 
- [ ] Security Alert fix 
  - [ ] Package update 
    - [ ] Major version update 
- [ ] Documentation 
- [ ] Workflow 
- [ ] Tests **only** 

## 4. Additional information

- [x] [Sprintlog](../SPRINTLOG.md) 
- [ ] Blocking PRs 
  - [ ] Merged 
- [ ] PR to `master` branch: \_If checked, read [the release instructions](../doc/procedures/new_release.md) 
  - [ ] I have followed steps 1-8. 

## 5. Actions / Scans

_Check the boxes when the specified checks have passed._

**For information on what the different checks do and how to fix it if they're failing, enter edit mode of this description or go to the [PR template](../.github/pull_request_template.md).**

- [x] **Black**

- [x] **Prettier**

- [x] **Yamllint**

- [x] **Tests**

- [x] **CodeQL**

- [x] **Trivy**

- [x] **Snyk**

### CODE REVIEW

**Review by i-oden — CHANGES_REQUESTED** [review]
Just a quick note -- add a description to the PR

**Comment by valyo** [comment]
Also, the title says "instead", and it should be "Alert unit and us..." or something like that

File: SPRINTLOG.md
Comment (i-oden) [inline SPRINTLOG.md:450]: Not instead, both

**Review by valyo — APPROVED** [review]

File: dds_web/commands.py
Comment (i-oden) [inline dds_web/commands.py:1310]:
```suggestion
                "Your unit is approaching the allocated data quota (see details below).\n\n"
                f"NB! If you would like to increase or decrease the allocated quota ('Quota') or the level after which you receive this email ('Warning level'), the technical contact person for your unit must send a request to {dds_contact}.\n"
```

**Review by i-oden — CHANGES_REQUESTED** [review]
Aside from my comments, we need to have some way of not spamming the units with these emails. Should we have a new table keeping track of if this email has been sent and in that case don't send this again (could probably add a reminder though, not in this PR though)? Any other / better ideas for keeping track of this? 
@ScilifelabDataCentre/teamhermes

**Comment by valyo** [comment]
> Aside from my comments, we need to have some way of not spamming the units with these emails. Should we have a new table keeping track of if this email has been sent and in that case don't send this again (could probably add a reminder though, not in this PR though)? Any other / better ideas for keeping track of this? @ScilifelabDataCentre/teamhermes

I am not sure I understand the question, what should we keep track of?

**Comment by valyo** [comment]
Wouldn't it be sent once a day, when the daily cronjob runs? In such case, if nothing has changed, i.e. threshold hasn't been changed and or usage hasn't gone down, then the email should be send again to all concerned parties.

**Comment by i-oden** [comment]
> Wouldn't it be sent once a day, when the daily cronjob runs? In such case, if nothing has changed, i.e. threshold hasn't been changed and or usage hasn't gone down, then the email should be send again to all concerned parties.

But the units have a threshold of 80%. I don't think it's reasonable to ask them to immediately message us to ask for a change in warning level, or to expect that we would be able to change the warning level for them immediately (worst case multiple units on the same day), and even if they reach the threshold they may not want to increase their quota because they may know that they won't be uploading that much now or that there are some projects that will be archived any second. We shouldn't force the units to change anything, and in the case that they would want to change the quota: We don't do that within a day, there are steps, partly within database but also to check with SUNET/Safespring that there is enough space, and to have them change settings, and to inform e.g. Jonas or other DCMG about the change / sometimes get approval. 

My first though was to  
- create a table `WarningEmails` (or something) with the columns `unit`, `date` and `active` 
- before sending this email, check if there's a row in the table 
- if there is a row in the table --> check when the row was added (when the email was sent) 
    - if the row was added less than a week ago --> don't send another email 
    - if the row was added at least a week ago --> send a reminder with 1. updated information, 2. info that they will keep getting this reminder (or something like this)
- if there's not a row in the table --> send email and add row to table 

Now thinking about it, I have another couple of options which might be better:
**Option 1** 
- May need a new table here too to keep track of which warning has been sent
- Change the `warning_level` column to a list or something if possible, to contain multiple levels, e.g. 80, 90, 95. 
- Cronjob does normal usage calculation and sees if any of the warning levels have been reached 
- In that case checks if the warning email has already been sent 
    - If the warning has already been sent --> check which level it was sent on and possibly send a new warning if new warning level has been passed 
    - If warning has not been sent --> send email
- If warning level has not been passed and there is a row in table --> remove rows from table 

**Option 2**
- Add columns to `Unit` table: `warnings_active` (bool) and `warnings_deactivated` (date and time) 
- When cronjob is run, check if `warnings_active == True` and in that case send the warning
    - Also include in warning email that if they do not message us to say they don't want warnings then they will keep getting them every day
- If `warnings_active == False` --> check if the warning has been deactivated for x number of days 
    - If so --> set `warnings_active` to True again and null `warnings_deactivated`
    - If not --> no warning sent 

@ScilifelabDataCentre/teamhermes These are just options and thoughts on my end and some are probably a bit too complicated / too much work, I'm completely open to suggestions.

**Comment by valyo** [comment]
My opinion is that such kind of warning should be as straightforward as possible. Trying to implement some logic in order to spare the units email notification when they are above the threshold is not going to do any good to them and/or to us. On the contrary, it will potentially lead to more work for us trying to handle it if they "dangerously" approach the quota, plus that it will complicate the system even more. 
I think that improving the system stability, which will allow us to onboard more units is far more important.

**Comment by rv0lt** [comment]
My commnets:

1. Agree about the contact email thing, didn't realise it could be sent to null, and that does not make sense. So agree it should be non nullable

2. About the emails, maybe we can discuss the periodicity. But I definitely think that this "spamming" is not bad. Most other cloud providers will also do that (attach some monthly emails I was getting from AWS). And actually, the thing is that if they get issues with the storage they will also complain, it is a way to cover ourselves: "Hey, you were getting alerts quite often, you should have done something". 
I don't see it as spam, as they agree to get alerts in that threshold. They don't want it? Okay, then change the **warning_level**


image

**Comment by i-oden** [comment]
> My opinion is that such kind of warning should be as straightforward as possible. Trying to implement some logic in order to spare the units email notification when they are above the threshold is not going to do any good to them and/or to us. On the contrary, it will potentially lead to more work for us trying to handle it if they "dangerously" approach the quota, plus that it will complicate the system even more. I think that improving the system stability, which will allow us to onboard more units is far more important.

@valyo
- Re: as straight forward as possible -- I completely agree. My thoughts / options were really just that, I agree that they would take too much time and effort atm. If we do fix something then it should be a simple fix.
- Re: that improving the system stability is far more important, ofc, I agree.

@rv0lt
- Re: 1 -- :+1::skin-tone-2:
- Re: your second point, yes, I agree, fair enough.

@ScilifelabDataCentre/teamhermes
My point was mainly: Yes, spamming them with these warnings is ok, until we know that they have the information. What if they tell us "yes, we want to change the quota", but we can't do that in a day or maybe even a week, due to contracts or general negotiations -- Do they (and we) just continue getting these warning emails every day until it's potentially fixed? This will lead us to get both the cronjob emails, and heaps of emails from the units about confusion "I thought we were fixing this already" and "please stop with these emails now" which will also take up our time.
I do agree though that they could just change the warning level if they don't want the warnings, BUT then we need to add this to the email. So then the message should be this in that case.
```python
message: str = (
    "Your unit is approaching the allocated data quota (see details below).\n\n"
    "NB! You will continue to get this email until you adjust your allocated quota ('Quota') "
    "or / and the level after which you receive this email ('Warning level'). "
    f"In order to change these settings, the technical contact person for your unit must send a request to {dds_contact}.\n"
    f"Unit name: {unit.name}"
    f"{info_string}"
)
```

File: migrations/versions/e02fe8fde71e_unit_contact_email_non_nullable.py
Comment (i-oden) [inline migrations/versions/e02fe8fde71e_unit_contact_email_non_nullable.py:20]: In the upgrade, also add values to rows where the contact_email is not filled in. See other migrations where similar things have been done.

**Review by i-oden — CHANGES_REQUESTED** [review]

**Review by i-oden — CHANGES_REQUESTED** [review]

File: migrations/README
Comment (rv0lt) [reply migrations/README:1]: The link you mentioned in the reply is just a typo right?

Comment (rv0lt) [reply migrations/README:1]: Actually I didn't know about this section in the root, I thought there was no info

Comment (i-oden) [reply migrations/README:1]: > The link you mentioned in the reply is just a typo right?

Oops, now the link is correct

Comment (i-oden) [reply migrations/README:1]: > Actually I didn't know about this section in the root, I thought there was no info

Aha, sorry about that, I didn't get that you didn't know about that section.

**Comment by rv0lt** [comment]
Checked the two instances and there is no current unit that has a null contact_email

**Review by i-oden — CHANGES_REQUESTED** [review]
Even though we currently don't have any values that are null, we still need to include the possibility in the migration. An example is seen in `666003748d14_change_active_nullable_and_default` for example, but there are others too (not regarding nullable=False though I don't think), add something similar.

File: migrations/versions/0cd0a3b251e0_unit_contact_email_non_nullable.py
Comment (i-oden) [inline migrations/versions/0cd0a3b251e0_unit_contact_email_non_nullable.py:24]: Currently other migrations e.g. `666003748d14_change_active_nullable_and_default.py` are using another method of executing these things, and I don't want to introduce different methods, so my suggestion is to change to the following
```suggestion
    unit_table = sa.sql.table("units", sa.sql.column("contact_email", mysql.TINYINT(display_width=1)))
    op.execute(unit_table.update().where(unit_table.c.contact_email == None).values(contact_email="delivery@scilifelab.se"))
```
If there rest of the team @ScilifelabDataCentre/teamhermes disagrees and thinks we should have SQL instead, then I want to change the other migrations to have this structure too.

**Review by i-oden — CHANGES_REQUESTED** [review]

File: migrations/versions/0cd0a3b251e0_unit_contact_email_non_nullable.py
Comment (rv0lt) [reply migrations/versions/0cd0a3b251e0_unit_contact_email_non_nullable.py:24]: I can change, it is esentially the same so i dont mind

Comment (valyo) [reply migrations/versions/0cd0a3b251e0_unit_contact_email_non_nullable.py:24]: I think it makes sense to be consistent, i.e. to use the `sa` (despite of SQL being easier for me to read because I am not that used to `sa` :) )

**Review by i-oden — APPROVED** [review]
OK to merge **once you have tested it locally** (using `flask db downgrade` and `upgrade` etc)

**Comment by rv0lt** [comment]
> OK to merge **once you have tested it locally** (using `flask db downgrade` and `upgrade` etc)

Tested manually

**Pull request merged**
